library/Sexp/src/Juvix/Sexp.hs

```diff
@@ -51,103 +57,105 @@ import Prelude (error)
-      | predBind name -> bindCase
-      where
-        changeCase xs = do
-          f atom xs (\xs -> foldSearchPredManualRecurse xs p1 p2)
-        -- G takes the computation, as its changes are scoped over the
-        -- calls.
-        bindCase =
-          Cons a <$> g atom xs (\xs -> foldSearchPredManualRecurse xs p1 p2)
-    Cons cs xs ->
-      Cons <$> foldSearchPredManualRecurse cs p1 p2 <*> foldSearchPredManualRecurse xs p1 p2
-    Nil -> pure Nil
+    Cons (Atom (A name _)) _
+      | predChange name ->
+        f t recurseDown
+    Cons {} -> traverse recurseDown t
+    Nil {} -> pure Nil
     Atom a
-      | predChange ":atom" -> do
-        f a (Atom a) (\xs -> foldSearchPredManualRecurse xs p1 p2)
-      | otherwise -> pure $ Atom a
-
--- | @foldPred@ searches the sexp structure given to it with a given
--- predicate. This predicate is ran on the @car@ of every list
+      | onAtom opt -> f t recurseDown
+      | otherwise -> pure (Atom a)
+  where
+    recurseDown sexp = traversePredOptStar sexp predChange f opt
+
+-- | @traversePredStar@ is just like @traversePredOptStar@ except the
+-- empty option set is given, meaning that the function will not match
+-- on atoms.
+traversePredStar ::
+  Monad f => B a -> (NameSymbol.T -> Bool) -> (B a -> (B a -> f (B a)) -> f (B a)) -> f (B a)
+traversePredStar t predChange automaticF =
+  traversePredOptStar t predChange automaticF mempty
+
+autoRecurse ::
+  Monad m => (B a -> m (B a)) -> B a -> (B a -> m (B a)) -> m (B a)
+autoRecurse automaticF xs recurse = do
+  newCons <- automaticF xs
+  case newCons of
+    Cons {} -> traverse recurse newCons
+    _______ -> pure newCons
+
+-- | @mapPredStar@ searches the sexp structure given to it with a
+-- given predicate. This predicate is ran on the @car@ of every list
 -- structure. This simulates searches for the head of a function
 -- call. When the predicate returns true, the function passed to
--- foldPred (which we shall refer to as f), is then called on the
+-- @mapPredStar@ (which we shall refer to as f), is then called on the
 -- @car@, and the @cdr@ of the list, giving back a new sexp
 -- structure. This new sexp structure is then recursed upon by
--- foldPred. NOTE that this does mean the structure handed back by f.
-foldPred :: B a -> (NameSymbol.T -> Bool) -> (Atom a -> B a -> B a) -> B a
-foldPred t pred f =
+-- @mapPredStar@. NOTE that this does mean the structure handed back
+-- by f.
+mapPredStar :: B a -> (NameSymbol.T -> Bool) -> (B a -> B a) -> B a
+mapPredStar t pred f =
   case t of
-    Cons (Atom atom@(A name _)) xs
-      | pred name ->
-        foldPred (f atom xs) pred f
-    Cons cs xs ->
-      Cons (foldPred cs pred f) (foldPred xs pred f)
+    Cons (Atom (A name _)) _
+      | pred name -> mapPredStar (f t) pred f
+    Cons _ _ -> map (\t -> mapPredStar t pred f) t
+    Atom ato -> Atom ato
     Nil -> Nil
-    Atom a -> Atom a
+
+traverse :: Monad m => (B a -> m (B b)) -> B a -> m (B b)
```

Comment:
Wouldn't it be better to have instances of `Traversable`, `Foldable`, `Functor`, etc?

Reply:
These aren't the functions over the generic itself, thus the instance of `Functor` `Traversable` etc would be misleading (The functor over a tree goes on every element, recursing down, where as this nested structure will keep the similar list style traversal).